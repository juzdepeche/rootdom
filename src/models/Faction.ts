import images from '../../assets/images';

export class Faction {
	public name: string;
	public value: number;
	public color: string;
	public image;

	constructor(name: string, value: number, color: string, image?) {
		this.name = name;
		this.value = value;
		this.color = color;
		this.image = image;
	}
}

export const Factions = [
	new Faction('Marquise de Cat', 10, '#FD5200', images.cat),
	new Faction('Underground Duchy', 8, '#D7A872', images.taupe),
	new Faction('Eyrie Dynasties', 7, '#4774B7', images.bird),
	new Faction('Vagabond', 5, '#FCFEFD', images.vagabond),
	new Faction('Riverfolk Company', 5, '#61DDFB', images.otter),
	new Faction('Woodland Alliance', 3, '#59BA53', images.mouse),
	new Faction('Corvid Consipiracy', 3, '#290D7E', images.raven),
	new Faction('Lizard Cult', 2, '#f1f435', images.lizard)
];

export const GetViableGameSum = (playerCount:number) => {
	switch (playerCount) {
		case 2:
			return 17;
		case 3:
			return 18;
		case 4:
			return 21;
		case 5:
			return 25;
		case 6:
			return 28;
		default:
			return 0;
	}
}

export const GetBigFactionsSum = (factions:Faction[], playerCount:number) => {
	let sum = 0;
	factions.sort((a, b) => a.value > b.value ? -1 : 1)
	for (let i = 0; i < playerCount; i++) {
		sum += factions[i].value;
	}
	return sum;
}