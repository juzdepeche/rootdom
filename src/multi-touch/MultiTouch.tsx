import React, { Component } from 'react';
import { Text, View, StyleSheet, Vibration } from 'react-native';
import { MultiTouchView } from 'expo-multi-touch';
import { ParticipantTarget, ParticipantTargetProps } from '../participant/ParticipantTarget';
import { Faction } from '../../App';

const SHOW_TIME = 4000;

class Participant {
	public ready = false;
	public faction:Faction;
	public ref;

	constructor () {
		this.ref = React.createRef();
	}
}

export default class MultiTouch extends Component<{factions:Faction[]}> {

	get ViableGameSums() {
		const sums = [];
		sums[2] = 17;
		sums[3] = 18;
		sums[4] = 21;
		sums[5] = 25;
		sums[6] = 28;
		return sums;
	}

	state = {
	//https://stackoverflow.com/a/38378350
		touches: {},
		dispatched: false
	};

	participants:Participant[] = [];
	dispatchedParticipantsCount = 0;  
	showing = false;
	
  	touchProps = {
		onTouchBegan: event => {
			if (this.showing) return;

			this.participants = this.participants.filter(participant => !!participant);
			
			if (this.state.dispatched || this.participants.length >= this.props.factions.length) return;

			const { identifier } = event;
			this.setState(previous => ({
				touches: {
					...previous.touches,
					[identifier]: event,
				}
			}));
			this.participants[identifier] = new Participant();
		},
		onTouchMoved: event => {
			if (this.showing) return;

			const { identifier } = event;
			if (!this.participants[identifier]) return;
			this.setState(previous => ({
				touches: {
				...previous.touches,
				[identifier]: event,
				},
			}));
		},
		onTouchEnded: event => {
			if (this.showing) return;

			const { identifier, deltaX, deltaY, isTap } = event;
			
			this.participants = this.participants.filter(participant => !!participant);
			if (this.state.dispatched && !this.participants[identifier]) return;

			if (this.state.dispatched) {
				this.dispatchedParticipantsCount--;
				if (this.dispatchedParticipantsCount <= 0) {
					this.showFactions();
				}
				return;
			}

			if (!this.participants[identifier]) return;
			// this.participants[identifier].ref.current.shrink(this.resetParticipantTarget);
			this.resetParticipantTarget(identifier);
		},
		onTouchCancelled: event => {
			if (this.showing) return;

			const { identifier, deltaX, deltaY, isTap } = event;

			this.participants = this.participants.filter(participant => !!participant);
			if (this.state.dispatched && !this.participants[identifier]) return;
			
			if (this.state.dispatched) {
				this.dispatchedParticipantsCount--;
				if (this.dispatchedParticipantsCount <= 0) {
					this.showFactions();
				}
				return;
			}

			if (!this.participants[identifier]) return;
			// this.participants[identifier].ref.current.shrink(this.resetParticipantTarget);
			this.resetParticipantTarget(identifier);
		},
	};

	showFactions = () => {
		this.showing = true;
		setTimeout(() => {
			const keys = this.participants.keys();
			for (const key of keys) {
				this.participants[key].ref.current.shrink(this.deleteParticipant);
			}
			this.showing = false;
			this.setState({dispatched: false});
		}, SHOW_TIME);
	}

	participantReady = (index) => {
		if (!this.participants[index]) return;
		this.participants[index].ready = true;
		this.dispatchRoles();
	}

  	dispatchRoles = () => {
		this.participants = this.participants.filter(participant => !!participant);

		if (this.participants.length > 1 && this.participants.every(p => p.ready))
		{
			Vibration.vibrate(100);
			const factions = this.getFactions(this.participants.length);
			this.participants.forEach(participant => {
				participant.faction = factions.pop();
			});
			this.setState({dispatched: true});
			this.dispatchedParticipantsCount = this.participants.length;
		}
	}

	getFactions = (playerCount) => {
		let factionValueSum = 0;
		let factions = [];
		const minimumValueSum = this.ViableGameSums[playerCount];

		let allFactionsValueSum = 0;
		let isImpossibleFactionValueSum = false;
		this.props.factions.forEach(faction => {
			allFactionsValueSum += faction.value
		});
		isImpossibleFactionValueSum = allFactionsValueSum < minimumValueSum;

		do {
			factionValueSum = 0
			factions = this.shuffleFactions();
	
			for (let i = 0; i < playerCount; i++) {
				factionValueSum += factions[i].value;
			}

		} while (factionValueSum <= minimumValueSum && !isImpossibleFactionValueSum);
		console.log(isImpossibleFactionValueSum);
		return factions;
	}

	resetParticipantTarget = (identifier) => {
		this.setState(previous => ({
			touches: {
			...previous.touches,
			[identifier]: null,
			},
		}));
		this.setState({dispatched: false});
		// delete this.participants[identifier];
	}

	private shuffleFactions = () => {
		const { factions } = this.props;
		const roles:Faction[] = [];
		factions.forEach(faction => {
			roles.push(faction);
		});
		
		const max = roles.length
		for (let i = 0; i < roles.length - 1; i++) {
			let randomIndex = Math.floor((Math.random() * (max - i) ) + i);
			const tempItem = roles[randomIndex];
            roles[randomIndex] = roles[i];
            roles[i] = tempItem;
		}
		
		return roles;
	}

	deleteParticipant = (index) => {
		delete this.participants[index];
		this.setState(previous => ({
			touches: {
			...previous.touches,
			[index]: null,
			},
		}));
	}

  	render() {
		const { touches, dispatched } = this.state;
		return (
		<View style={{ flex: 1 }}>
			<MultiTouchView style={{ flex: 1 }} {...this.touchProps}>
				<View style={styles.container}>
					{Object.values(touches).map((item, index) => {
						if (!item || !this.participants[index]) {
							return null;
						}
						return (
							<ParticipantTarget 
								key={index} 
								ref={this.participants[index].ref}
								item={item} 
								index={index} 
								onReadyCallback={this.participantReady} 
								dispatched={dispatched}
								faction={this.participants[index].faction}
							/>
						)
					})}
				</View>
			</MultiTouchView>
		</View>
		);
  	}
}

const TOUCH_SIZE = 100;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000000',
  },
  touch: {
    position: 'absolute',
    aspectRatio: 1,
    width: TOUCH_SIZE,
    borderRadius: TOUCH_SIZE / 2,
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
    color: '#34495e',
  },
});