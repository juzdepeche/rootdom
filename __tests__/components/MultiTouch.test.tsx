import React from 'react';
import renderer from 'react-test-renderer';
import { Factions } from '../../src/models/Faction';
import MultiTouch from '../../src/components/MultiTouch';

describe('<MultiTouch />', () => {
	it('renders correctly', () => {
		const multiTouch = renderer
			.create(<MultiTouch factions={[...Factions]}></MultiTouch>)
			.toJSON();
		expect(multiTouch).toMatchSnapshot();
	});
});
