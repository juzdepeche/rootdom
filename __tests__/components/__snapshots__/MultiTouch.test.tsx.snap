// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiTouch /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onMoveShouldSetPanResponder={[Function]}
    onMoveShouldSetPanResponderCapture={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onPanResponderTerminationRequest={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onShouldBlockNativeResponder={[Function]}
    onStartShouldSetPanResponder={[Function]}
    onStartShouldSetPanResponderCapture={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchBegan={[Function]}
    onTouchCancelled={[Function]}
    onTouchEnded={[Function]}
    onTouchMoved={[Function]}
    onTouchesBegan={[Function]}
    onTouchesCancelled={[Function]}
    onTouchesEnded={[Function]}
    onTouchesMoved={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;
